### ISSUE TITLE
Allow using of numpad `,` 

### ISSUE THREAD

**Issue by heckflosse** [Beep6581/RawTherapee#6148]
Some years ago, I made a code change to allow behaving numpad `,` as `.`
This simplified entering fractional values on non english keyboards

https://github.com/Beep6581/RawTherapee/issues/2788#issuecomment-130070395

But somehow this change is gone...

I can't even find it in history as it was at google code times...

Anyone with an idea how to do that again or why it has gone?

**Comment by heckflosse** [comment]
I found the code. It's still there, but doesn't work anymore :(

https://github.com/Beep6581/RawTherapee/blob/dev/rtgui/guiutils.cc#L1214

**Comment by ff2000** [comment]
I just tried it on linux and it works as expected. Perhaps it's an issue with the windows version of GTK?
Does it help if you replace `return Gtk::Widget::on_key_press_event(event);` with `return Gtk::SpinButton::on_key_press_event(event);`?

I also wonder why Gtk::SpinButton isn't locale aware. The decimal point is `,` in German but the SpinButton insists on `.`? In Qt QDoubleSpinBox is locale aware, values are entered and shown as "1,234" and value() correctly transforms it to 1.234.

**Comment by heckflosse** [comment]
> Does it help if you replace return Gtk::Widget::on_key_press_event(event); with return Gtk::SpinButton::on_key_press_event(event);?

No :(

**Comment by Thanatomanic** [comment]
FWIW, the [Gdk documentation](https://developer.gnome.org/gdk3/unstable/gdk3-Event-Structures.html#GdkEventKey) says that using `event->string` is deprecated. 

> gchar *string; | a string containing an approximation of the text that would result from this keypress. The only correct way to handle text input of text is using input methods (see GtkIMContext), so this field is deprecated and should never be used. (gdk_unicode_to_keyval() provides a non-deprecated way of getting an approximate translation for a key.) The string is encoded in the encoding of the current locale (Note: this for backwards compatibility: strings in GTK+ and GDK are typically in UTF-8.) and NUL-terminated. In some cases, the translation of the key code will be a single NUL byte, in which case looking at length is necessary to distinguish it from the an empty translation.

And as for doing stuff with user input, the recommended way seems to be by using the `signal_input` method https://developer.gnome.org/gtkmm/3.24/classGtk_1_1SpinButton.html#ac1bffd7a579a68762e7470db52bb96fd
So what we try to do by catching a more low-level event, may be wrong. But I'm not sure... maybe it _should_ still work.

**Comment by heckflosse** [comment]
@Thanatomanic Roel, thanks for the hints. As I'm not a GUI coder I will let others try to solve that. Maybe @Pandagrapher or @Lawrence37 can chime in...

**Comment by ff2000** [comment]
And using signal_output we can also format the displayed number according to locale:
https://stackoverflow.com/questions/4827831/how-to-change-display-in-gtk-spinbutton

**Comment by Pandagrapher** [comment]
@heckflosse 
On MacOS, actual code works too. So I can't really test but this code works the same and at least doesn't use deprecated code:
```diff
diff --git a/rtgui/guiutils.cc b/rtgui/guiutils.cc
index 601fa42..04be410 100644
--- a/rtgui/guiutils.cc
+++ b/rtgui/guiutils.cc
@@ -1172,6 +1172,7 @@
     set_numeric(true);
     set_wrap(false);
     set_alignment(Gtk::ALIGN_END);
+    set_update_policy(Gtk::SpinButtonUpdatePolicy::UPDATE_IF_VALID); // Avoid updating text if input is not a numeric
 }
 
 void MySpinButton::updateSize()
@@ -1204,19 +1205,16 @@
     double vMin, vMax;
     get_range(vMin, vMax);
 
-    if ( (event->string[0] >= 'a' && event->string[0] <= 'z')
-            || (event->string[0] >= 'A' && event->string[0] <= 'Z')
-            || event->string[0] == '+' || (event->string[0] == '-' && vMin >= 0)
-            || event->string[0] == '=' || event->string[0] == '_'
-       ) {
-        return false;
+    if (event->keyval == GDK_KEY_plus || (event->keyval == GDK_KEY_minus && vMin >= 0)) {
+        return true; // Event is not propagated further
     } else {
-        if(event->string[0] == ',') {
-            event->keyval = GDK_KEY_period;
-            event->string[0] = '.';
+        if (event->keyval == GDK_KEY_comma) {
+            set_text(get_text() + ".");
+            set_position(get_text().length()); // When setting text, cursor position is reseted at text start. Avoiding this with this code
+            return true; // Event is not propagated further
         }
 
-        return Gtk::Widget::on_key_press_event(event);
+        return Gtk::SpinButton::on_key_press_event(event); // Event is propagated normally
     }
 }
```

Pierre

**Comment by heckflosse** [comment]
@Pandagrapher Pierre, I tried your patch, but unfortunately it doesn't work here :(

**Comment by Thanatomanic** [comment]
There is this funny line in `main.cc` that seems to force a decimal period. But unfortunately, commenting it out doesn't magically make it locale dependent.
https://github.com/Beep6581/RawTherapee/blob/2366cb4935165a86e86be3c608548a48bd48fa39/rtgui/main.cc#L369-L373

**Comment by heckflosse** [comment]
> There is this funny line in main.cc that seems to force a decimal period.

Yes, that was needed to fix a bug when reading and writing pp3 files. There is an issue about that somewhere on github.
It took us a lot of time to nail and fix it...

**Comment by Pandagrapher** [comment]
> @Pandagrapher Pierre, I tried you patch, but unfortunately it doesn't work here :(

Ok Gdk seems to have created specific `keyval` value for numpad key... In my patch, should try using this instead:
`if (event->keyval == GDK_KEY_comma || event->keyval == GDK_KEY_KP_Decimal) {`

Pierre

**Comment by heckflosse** [comment]
@Pandagrapher That works. \o/ Feel free to push :+1:

Edit: and of course: thank you :+1:

**Comment by heckflosse** [comment]
@Pandagrapher You can not know, how glad I am, that this issue, which annoyed me for very long time, can be fixed now. It's such a time saver to be able to use the numpad on a german keyboard again in RT when entering values :+1:

**Comment by Pandagrapher** [comment]
@heckflosse You're welcomed ;) Pushed

**Comment by heckflosse** [comment]
@Pandagrapher While `,` on numpad now works as `.`, the downside is, that pressing `f` or `z` when focus is on an adjuster value, does not work anymore (it does not zoom in or fit to screen)

**Comment by heckflosse** [comment]
@Pandagrapher 

Most likely because you changed this condition:

```
   if ( (event->string[0] >= 'a' && event->string[0] <= 'z')
           || (event->string[0] >= 'A' && event->string[0] <= 'Z')
           || event->string[0] == '+' || (event->string[0] == '-' && vMin >= 0)
           || event->string[0] == '=' || event->string[0] == '_'
      ) {
       return false;

```

**Comment by Pandagrapher** [comment]
@heckflosse 
Sorry... Thought that was some left useless code or bug. It's now fixed with commit `84751ad`

**Comment by heckflosse** [comment]
@Pandagrapher No need to apologize. Works fine now :+1:
